### PR TITLE
chore: add precommit hook

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Motivation: I found I frequently forgot fix typos/rustfmt when pushing changes. Checking the CI and fix it is both wasting CI resource and my time. Having a pre-commit hook is helpful to exclude some trivial errors. (And I don't want to run `risedev c` because clippy is slow..)

Note that this is not mandatory but opt-in.

Alternatively we can do something like auto push a fix commit in CI, but some people may not like it. :p

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by kwannoel — APPROVED** [review]

**Comment by BugenZhao** [comment]
https://github.com/risingwavelabs/risingwave-legacy/pull/2704

**Comment by BugenZhao** [comment]
> [risingwavelabs/risingwave-legacy#2704](https://github.com/risingwavelabs/risingwave-legacy/pull/2704)

Well, it's somewhat different. The previous `rusty_hook` will automatically be enabled for all developers, while `pre-commit` is totally optional.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+-   repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v2.3.0
+    hooks:
+    -   id: end-of-file-fixer
+    -   id: trailing-whitespace
+-   repo: local
+    hooks:
+    -   id: rustfmt
+        name: rustfmt
+        entry: rustfmt
```
Comment (BugenZhao) [inline .pre-commit-config.yaml:14]: - `rustfmt` vs `cargo fmt`?
- `fmt` vs `fmt --check`?

Comment (xxchan) [reply .pre-commit-config.yaml:14]: It will run against diff, so rustfmt. And it is silly to check but not fix.

**Pull request merged**